graphene_django/debug/middleware.py

```diff
@@ -67,3 +70,28 @@ def resolve(self, next, root, info, **args):
             return context.django_debug.on_resolve_error(e)
         context.django_debug.add_result(result)
         return result
+
+
+class DjangoSyncRequiredMiddleware:
+    def resolve(self, next, root, info, **args):
+        parent_type = info.parent_type
+        return_type = info.return_type
+
+        if isinstance(parent_type, GraphQLNonNull):
+            parent_type = parent_type.of_type
+        if isinstance(return_type, GraphQLNonNull):
+            return_type = return_type.of_type
+
+        if any(
+            [
+                hasattr(parent_type, "graphene_type")
+                and hasattr(parent_type.graphene_type._meta, "model"),
+                hasattr(return_type, "graphene_type")
+                and hasattr(return_type.graphene_type._meta, "model"),
+                info.parent_type.name == "Mutation",
+            ]
+        ):
+            if is_sync_function(next):
+                return sync_to_async(next)(root, info, **args)
```

Comment:
Based on what I understand from the [documentation](https://github.com/django/asgiref), this is going to wrap any synchronous resolver that accesses the DB into a separate thread, isn't it?

If this is the case indeed, isn't this going to have a substantial bad impact on performances?

Reply:
For CPU bound resolvers the overhead will make the performance a lot worse. Since most Django apps are backed by an external DB, most resolvers are IO bound - so the advantage comes from the main loop can continue to serve other requests/continue resolving other parts of the query.

Reply:
agree with @jaw9c, also this is the main approach of Django core team on the way to making Django async

Reply:
While I agree most resolvers are IO bound, we still miss benchmarks to assert that performance will improve with this pattern. The cost of creating a thread for each resolver + event loop task context switching VS blocking DB calls tradeoff is not obvious at all and I wouldn't be surprised if the synchronous version performs better for simple queries.

Reply:
no, it still makes sense to support `sync_to_async` mode for those projects that need async support. Benchmarks for sync resolvers run under `AsyncGraphQLView` (ASGI) will be definitely worse than sync resolvers under  `GraphQLView` (WSGI) as you said.

But this PR is for projects with async resolvers running under `AsyncGraphQLView` (ASGI) - there is no possibility to run such code without this PR now

Reply:
@jaw9c It sounds like you have a way to get some sort of benchmarks to show the performance impact, but if not I can help add those. I'm not sure what the best way would be to get them, but I can do my best :smile:

Reply:
@superlevure but what is the alternative to use for dataloaders ? In the documentation there are only async dataloaders mentioned.

Dataloaders are an important part of any graphql API, I don't understand how the upgrade was done to v3 without taking into consideration dataloaders?!

There is the sync dataloaders package(https://github.com/jkimbo/graphql-sync-dataloaders) but that is an external package, what solution did you guys have in mind for dataloaders in the new graphene versions?

Reply:
Agreed, the current situation regarding dataloaders is unfortunate. We've been working on adding them back on v3 thanks to `graphql-sync-dataloaders` on our fork of `graphene-django` at my company. You can check out our tentative here:

- https://github.com/loft-orbital/graphene-django/pull/37
- https://github.com/loft-orbital/graphene-django/pull/44

We've been using this in production for a few months now with success, our plan is to eventually submit this to upstream when we are confident enough and we have the bandwidth.

Reply:
Note that our implementation of dataloaders for many-to-one relations is not covered in those two PRs as it's implemented in a private repo, but it basically looks like this:

```python
    @classmethod
    def get_node(cls, info, id):
        if not graphene_settings.USE_DATALOADERS:
            return super().get_node(info, id)

        model_name: str = cls._meta.model._meta.model_name
        queryset: QuerySet = cls.get_queryset(cls._meta.model.objects, info)

        if not hasattr(info.context, "dataloaders"):
            info.context.dataloaders = {}

        if model_name not in info.context.dataloaders:

            def load_many(keys: list[UUID | str]):
                qs: QuerySet = queryset.filter(id__in=keys)
                object_map: dict[str, Any] = {str(object_.id): object_ for object_ in qs}

                return [object_map.get(str(key), None) for key in keys]

            info.context.dataloaders[model_name] = SyncDataLoader(load_many)

        return info.context.dataloaders[model_name].load(id)
```

Reply:
I did not realise that for many-to-one relations you actually need another fix to get this working.
At this point, if someone does not fix dataloaders, I will just start using another graphql library because its getting ridiculous.